addons/website_sale/models/website.py

```diff
@@ -60,35 +60,53 @@ def _get_pl_partner_order(self, country_code, show_visible, website_pl, current_
+        #          does not invalidate the cache. Pass it through `partner_pl` parameter instead.
+
         self.ensure_one()
         pricelists = self.env['product.pricelist']
         if country_code:
             for cgroup in self.env['res.country.group'].search([('country_ids.code', '=', country_code)]):
-                for group_pricelists in cgroup.pricelist_ids.filtered(lambda pl: pl.website_id.id in (False, self.id)):
-                    if not show_visible or group_pricelists.selectable or group_pricelists.id in (current_pl, order_pl):
-                        pricelists |= group_pricelists
+                pricelists |= cgroup.pricelist_ids.filtered(lambda pl: pl._is_available_on_website(self.id) and _check_show_visible(pl))
 
-        partner = self.env.user.partner_id.sudo(user=self.env.user)
         is_public = self.user_id.id == self.env.user.id
-        if not is_public and (not pricelists or (partner_pl or partner.property_product_pricelist.id) != website_pl):
-            # `property_product_pricelist` is already multi-website compliant
-            pricelists |= partner.property_product_pricelist
-
-        if not pricelists:  # no pricelist for this country, or no GeoIP
-            pricelists |= all_pl.filtered(lambda pl: not show_visible or pl.selectable or pl.id in (current_pl, order_pl))
-        if not show_visible and not country_code:
-            pricelists |= all_pl.filtered(lambda pl: pl.sudo().code)
+        # If the user is logged in and either he has no country pricelists or either
+        # his partner pricelist is not the website pricelist, then we use his partner pricelist
+        if not is_public and (not pricelists or partner_pl != website_pl):
+            pricelists |= pricelists.browse(partner_pl)
+
+        if is_public and not pricelists:  # no pricelist for this country, or no GeoIP
+            # TODO (review) if is_public was implicit as you cant go in this is if you are not public given the previous if
+            pricelists |= all_pl.filtered(lambda pl: _check_show_visible(pl))
+        if not country_code:
+            # TODO (review) this if condition as always been weirdoooo
+            # Introduced with https://github.com/odoo/odoo/commit/60300fc287928358d31ae5faff1b7f9953c57d18
+            # to allow logged in user to be able to use promo code but it should actually have been this directly, not just filtered on code
+            # Then fixed with https://github.com/odoo/odoo/commit/e11908d429435091828f224ee0751b90de8c99aa
+            # to allow user to see selectable pricelists but the fix makes no sense as it still only filter on code..
+            pricelists |= all_pl.filtered(lambda pl: _check_show_visible(pl))
 
         # This method is cached, must not return records! See also #8795
-        return pricelists.filtered(lambda pl: pl._is_available_on_website(self.id)).ids
+        return pricelists.ids
```

Comment:
can we simplify with:


```python
if country_code:
    for cgroup in self.env['res.country.group'].search([('country_ids.code', '=', country_code)]):
        pricelists |= cgroup.pricelist_ids.filtered(lambda pl: pl._is_available_on_website(self.id) and _check_show_visible(pl))
if not country_code or not pricelists:
    pricelists |= all_pl.filtered(lambda pl: _check_show_visible(pl))


is_public = self.user_id.id == self.env.user.id
if not is_public:
    pricelists |= pricelists.browse(partner_pl)
```

 + need

```python
# add current pl (e.g. if promo pl, country_code not set on pricelist (ignored btw), but set in method)
pricelists |= pricelists.browse(current_pl)
```

Reply:
Done.
About
```python
# add current pl (e.g. if promo pl, country_code not set on pricelist (ignored btw), but set in method)
pricelists |= pricelists.browse(current_pl)
```
I added a test `test_get_pricelist_available_geoip4` that test that current_pl is added and it works without that line.

So I am probably not understanding your flow (public user? how can he set its current_pl into session if not shown in the first place?) let me know.